### ISSUE TITLE
Consequently Operation of BulkInsertOrUpdateOrDelete Issue

### ISSUE THREAD

**Issue by dineshhp82** [borisdj/EFCore.BulkExtensions#113]
Hi,
I have face issue in bulk operation insertorupdateordelete while performing operation in consequently manner. 
Suppose I have execute the BulkInsertOrUpdateOrDelete operation first time in db table(or say entity) which insert 100 rows and again perform same operation on same table for more 50 rows then It will break your code.
e.g if you do small change in your code for class EFCoreBulkTest and run RunInsertOrUpdate method two time then it will fail your test case.

public void OperationsTest(bool isBulkOperation)
        {
. . . . . . .
     RunInsertOrUpdate(isBulkOperation); 
            RunInsertOrUpdate(isBulkOperation);
. . . . . . . 
            CheckQueryCache();
        }

But in RunInsertOrUpdate I have used BulkInsertOrUpdateOrDelete operation.
**context.BulkInsertOrUpdateOrDelete(entities, bulkConfig, (a) => WriteProgress(a));**

Getting below error message.
_System.Data.SqlClient.SqlException: 'Cannot insert the value NULL into column 'ItemId', table 'EFCoreBulkTest.dbo.ItemTempf4bee1e9Output'; column does not allow nulls. INSERT fails.
The statement has been terminated.'_

I have few query related to bulk operation
.
1)In your test project there is an entity Item having ItemId is PK and Identity column then why we are assign the value to ItemId while creating collection of data. As this is already Identity column then It should
Auto increment and fill its values. If I look into your code then while you create the Temp(no Output Temp) table to copy the data for input collection then It have no identity either start from zero(SetOutputIdentity=false) Or 1,2,3 ( SetOutputIdentity=true). But if table have already data with same identity then it will conflict and code will break at merge operation.

I think for Temp table start value for ItemId(any PK Identity column) should be last Identity value of table. By using any sql query to find last identity Ident_Identity(‘’).

2)While we set PreserveInsertOrder=false the why you clear your list of items.Because you fill your collection entitiesWithOutputIdentity from FullTempTable not from Orginal Table.
It will lose your existing data.

public void LoadOutputData<T>(DbContext context, IList<T> entities) where T : class
        {
            if (BulkConfig.SetOutputIdentity && HasSinglePrimaryKey)
            {
                string sqlQuery = SqlQueryBuilder.SelectFromOutputTable(this);//========== **$"SELECT {GetCommaSeparatedColumns(columnsNames)} FROM {tableInfo.FullTempOutputTableName}";**
                var entitiesWithOutputIdentity = QueryOutputTable<T>(context, sqlQuery).ToList();
                UpdateEntitiesIdentity(entities, entitiesWithOutputIdentity);
            }

As entitiesWithOutputIdentity is getting data from FullTempOutTable not from original table
 
protected void UpdateEntitiesIdentity<T>(IList<T> entities, IList<T> entitiesWithOutputIdentity)
        {
            if (BulkConfig.PreserveInsertOrder) // Updates PK in entityList
            {
              **//If you preserve Insert Order then Your new entitiesWithOutputIdentity has same entity which is already exist on database.**
                var accessor = TypeAccessor.Create(typeof(T), true);
                for (int i = 0; i < NumberOfEntities; i++)
                    accessor[entities[i], PrimaryKeys[0]] = accessor[entitiesWithOutputIdentity[i], PrimaryKeys[0]];
            }
            else // Clears entityList and then refills it with loaded entites from Db
            {
                **entities.Clear();**
                ((List<T>)entities).AddRange(entitiesWithOutputIdentity);
            }
        }

Thanks In advance.  let me correct my understanding if I am wrong.

**Comment by borisdj** [comment]
You can call same Bulk operation several times, one after another.
If you want to test it you should make new test for that.
Existing test are made be to pass only if they called as they are now.
What issue(Exception) do you have and in what use case exactly?
For info on documentation read the README first detailed, not just looking at tests because for simplicity some are written in way that that is not standard usage.

**Comment by dineshhp82** [comment]
Gettting below exception
System.Data.SqlClient.SqlException: 'Cannot insert the value NULL into column 'ItemId', table 'EFCoreBulkTest.dbo.ItemTempf4bee1e9Output'; column does not allow nulls. INSERT fails.
The statement has been terminated.'

**Comment by borisdj** [comment]
Regarding your questions:

1) You don't have to set _ItemId_, since it is Identity column it will be set by Db.
But if you want to keep Order than it should be set for mutual sorting, negative values can be used and DB will change them.

2) After Output result is return from DB to be able to update Id(s) of original data list I have used their position in the list so they need to be sorted for correct matching.
So if you want initial list to be updated then just use also _PreserveInsertOrder=true_.
Otherwise list will be cleared and reloaded, but you don't lose any data since all data were first written into Db and then just Selected back.

As for the Exception you have, I can confirm the issue.
When MERGE has 'NOT MATCHED BY SOURCE THEN DELETE'
and that combined with OUTPUT, in INSERTED list it will return all changes including ones that were Deleted with all NULL values and since Id column in TempOutput table is not Nullable it breaks.
[The output clause will output one row for every row affected by the MERGE.](https://social.msdn.microsoft.com/Forums/sqlserver/en-US/6ca621a8-35cf-40cb-8998-ac414c479450/only-show-insert-actions-on-output-of-merge?forum=transactsql)

This can be reproduced in SQL directly by using code from TestExample.sql in the following way:
```SQL
-- Clear Tables and Resets AutoIncrement
DELETE FROM [his].[ItemHistory];
DELETE FROM [dbo].[Item];
DBCC CHECKIDENT ('[dbo].[Item]', RESEED, 0);

-- CREATE TempTable
SELECT TOP 0 T.[ItemId], T.[Description], T.[Name], T.[Price], T.[Quantity], T.[TimeUpdated]
INTO [dbo].[ItemTemp1234] FROM [dbo].[Item] AS T
LEFT JOIN [dbo].[Item] AS Source ON 1 = 0; -- removes Identity constrain
-- CREATE TempTableOutput
SELECT TOP 0 T.[ItemId], T.[Description], T.[Name], T.[Price], T.[Quantity], T.[TimeUpdated]
INTO [dbo].[ItemTemp1234Output] FROM [dbo].[Item] AS T
LEFT JOIN [dbo].[Item] AS Source ON 1 = 0;

-- INSERT INTO TempTable
INSERT INTO [dbo].[ItemTemp1234]
([ItemId], [Description], [Name], [Price], [Quantity], [TimeUpdated])
VALUES
(-3, 'Desc1', 'SomeName1', 22.11, 34, '2017-01-01'),
(-2, 'Desc2', 'SomeName2', 12.66, 45, '2017-01-01'),
(-1, 'Desc3', 'SomeName3', 14.35, 46, '2017-01-01');
```
Now do the initial Merge
```SQL
-- INSERT/UPDATE(DELETE) with MERGE from TempTable
MERGE [dbo].[Item] WITH (HOLDLOCK) AS T
USING (SELECT TOP 2147483647 * FROM [dbo].[ItemTemp1234] ORDER BY [ItemId]) AS S
ON T.[ItemId] = S.[ItemId]
WHEN NOT MATCHED BY TARGET THEN INSERT ([Description], [Name], [Price], [Quantity], [TimeUpdated])
VALUES (S.[Description], S.[Name], S.[Price], S.[Quantity], S.[TimeUpdated])
WHEN MATCHED THEN UPDATE SET T.[Description] = S.[Description], T.[Name] = S.[Name], T.[Price] = S.[Price], T.[Quantity] = S.[Quantity], T.[TimeUpdated] = S.[TimeUpdated]
WHEN NOT MATCHED BY SOURCE THEN DELETE
OUTPUT INSERTED.[ItemId], INSERTED.[Description], INSERTED.[Name], INSERTED.[Price], INSERTED.[Quantity], INSERTED.[TimeUpdated]
INTO [dbo].[ItemTemp1234Output];
```

Next clear the TempOutput table
```SQL
DELETE FROM [dbo].[ItemTemp1234Output];
```

Then execute the Merge segment again and it will break:

> Cannot insert the value NULL into column 'ItemId', table 'EFCoreBulkTest.dbo.ItemTemp1234Output'; column does not allow nulls. INSERT fails.

If last line (INTO [dbo].[ItemTemp1234Output]) from Merge would be removed in the Result window we get:
```
7	Desc1	SomeName1	22.11	34	2017-01-01 00:00:00.0000000
8	Desc2	SomeName2	12.66	45	2017-01-01 00:00:00.0000000
9	Desc3	SomeName3	14.35	46	2017-01-01 00:00:00.0000000
NULL	NULL	NULL	NULL	NULL	NULL
NULL	NULL	NULL	NULL	NULL	NULL
NULL	NULL	NULL	NULL	NULL	NULL
```

I haven't yet found any way to Filter out these Nulls from Output.
If that can't be done other possible option could be to change columns TempOutput table so they are all Nullable and then on Selecting them omit the Nulls.

For now you can use Sync without SetOutputIdentity, and then just use BulkRead after.

**Comment by borisdj** [comment]
Fixed with v3.4.0
Output table columns are nullable as proposed in previous comment.
